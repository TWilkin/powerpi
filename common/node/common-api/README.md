# PowerPi - `@powerpi/common-api` - Typescript Library

PowerPi Common API Typescript library provides all the shared code utilised by typescript services that call the _API_ in PowerPi.

The library is built using typescript, with dependencies using yarn workspaces.

## Building

The library can be built by executing the following commands.

```bash
# From the root of your PowerPi checkout
# Download the dependencies
yarn

# Build the common API library
yarn build:common-api
```

## Testing

There are currently no automated tests for this library.

## Local Execution

This library cannot be executed itself, its code is executed within the services that depend on it.
