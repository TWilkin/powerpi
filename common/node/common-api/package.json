{
    "name": "@powerpi/common-api",
    "version": "0.1.9",
    "description": "PowerPi Common API Typescript library",
    "type": "module",
    "private": true,
    "license": "GPL-3.0-only",
    "author": "TWilkin",
    "repository": {
        "type": "git",
        "url": "https://github.com/TWilkin/powerpi.git",
        "directory": "common/node/api"
    },
    "main": "dist/src/index.js",
    "types": "dist/src/index.d.ts",
    "scripts": {
        "build": "tsc"
    },
    "files": [
        "dist/src"
    ],
    "devDependencies": {
        "typescript": "^5.1.6"
    },
    "dependencies": {
        "axios": "^1.12.0",
        "http-status-codes": "^2.1.4",
        "socket.io-client": "^4.7.2"
    }
}
