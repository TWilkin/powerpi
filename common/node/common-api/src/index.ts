import AdditionalState from "./AdditionalState";
import BaseDevice from "./BaseDevice";
import Battery from "./Battery";
import { BatteryStatusMessage } from "./BatteryStatus";
import Capability from "./Capability";
import { CapabilityStatusCallback, CapabilityStatusMessage } from "./CapabilityStatus";
import Config from "./Config";
import { ConfigFileType, ConfigStatusCallback, ConfigStatusMessage } from "./ConfigStatus";
import Device from "./Device";
import DeviceChangeMessage from "./DeviceChangeMessage";
import DeviceState from "./DeviceState";
import { DeviceStatusCallback, DeviceStatusMessage } from "./DeviceStatus";
import { Floor, Floorplan, Point, PolygonRoom, RectangleRoom, Room } from "./Floorplan";
import History from "./History";
import { Metric, MetricValue } from "./Metric";
import PaginationResponse from "./Pagination";
import PowerPiApi from "./PowerPiApi";
import Sensor from "./Sensor";
import { SensorStatusCallback, SensorStatusMessage } from "./SensorStatus";
import SocketIONamespace from "./SocketIONamespace";

export {
    AdditionalState,
    BaseDevice,
    Battery,
    BatteryStatusMessage,
    Capability,
    CapabilityStatusCallback,
    CapabilityStatusMessage,
    Config,
    ConfigFileType,
    ConfigStatusCallback,
    ConfigStatusMessage,
    Device,
    DeviceChangeMessage,
    DeviceState,
    DeviceStatusCallback,
    DeviceStatusMessage,
    Floor,
    Floorplan,
    History,
    Metric,
    MetricValue,
    PaginationResponse,
    Point,
    PolygonRoom,
    PowerPiApi,
    RectangleRoom,
    Room,
    Sensor,
    SensorStatusCallback,
    SensorStatusMessage,
    SocketIONamespace,
};
