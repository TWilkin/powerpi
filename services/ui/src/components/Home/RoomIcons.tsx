import { Room as IRoom } from "@powerpi/common-api";
import { useCallback, useMemo } from "react";
import _ from "underscore";
import DeviceIcon from "../Components/DeviceIcon";
import SensorIcon from "../Components/SensorIcon";
import useRoomDevices from "./useRoomDevices";

type RoomIconsProps = {
    room: IRoom;
    rotate: boolean;
};

const RoomIcons = ({ room, rotate }: RoomIconsProps) => {
    const devices = useRoomDevices(room.name);

    console.log(rotate);

    const { iconSize, deviceCount, iconsWide, iconsTall, offsetX, offsetY } = useMemo(() => {
        const iconSize = 32;

        // calculate the room size
        let width = room.width ?? 1;
        let height = room.height ?? 1;

        // this could put the icons outside the polygon so needs to be improved
        if (room.points) {
            width =
                Math.max(...room.points.map((point) => point.x)) -
                Math.min(...room.points.map((point) => point.x));

            height =
                Math.max(...room.points.map((point) => point.y)) -
                Math.min(...room.points.map((point) => point.y));
        }

        // next we need to find the centre of the room
        const centreX = width / 2 + (room.x ?? 0);
        const centreY = height / 2 + (room.y ?? 0);

        console.log("centre", centreX, centreY);

        // how many icons can we fit?
        let iconsWide = Math.floor(width / iconSize);
        let iconsTall = Math.floor(height / iconSize);

        // we want to leave some space either side
        iconsWide = Math.max(iconsWide - 4, 0);
        iconsTall = Math.max(iconsTall - 4, 0);

        // now we know what the limits are, arrange what we want
        if (iconsWide >= devices.length) {
            iconsTall = 1;
        }
        const deviceCount = Math.min(devices.length, iconsWide * iconsTall);

        // now we can work out the offsets
        const offsetX = centreX - (iconsWide * iconSize) / 2;
        const offsetY = centreY - (iconsTall * iconSize) / 2;

        return {
            iconSize,
            deviceCount,
            iconsWide,
            iconsTall,
            offsetX,
            offsetY,
        };
    }, [devices.length, room.height, room.points, room.width, room.x, room.y]);

    console.log(room);
    console.log(room.name, deviceCount, iconsWide, "x", iconsTall, "offset", offsetX, ",", offsetY);

    const getTransform = useCallback(
        (type: string, index: number) => {
            // find out which row and column it's in
            let row = 0;
            while (index >= (row + 1) * iconsWide) {
                row++;
            }
            const column = index - row * iconsWide;

            // now we can work out the position
            const x = offsetX + column * iconSize;
            const y = offsetY + row * iconSize;

            console.log(type, index, ":", row, ",", column, " ", x, ",", y);

            const translate = `translate(${x}, ${y})`;

            if (rotate) {
                return `${translate} rotate(-90)`;
            }

            return translate;
        },
        [iconSize, iconsWide, offsetX, offsetY, rotate],
    );

    return (
        <>
            {_(devices)
                .first(deviceCount)
                .map((device, i) => (
                    <g
                        key={`${device.deviceType}_${device.type}`}
                        transform={getTransform(device.type, i)}
                    >
                        <title>
                            {device.type} ({device.count})
                        </title>

                        {device.deviceType === "device" && (
                            <DeviceIcon type={device.type} width={iconSize} height={iconSize} />
                        )}

                        {device.deviceType === "sensor" && (
                            <SensorIcon type={device.type} width={iconSize} height={iconSize} />
                        )}
                    </g>
                ))}
        </>
    );
};
export default RoomIcons;
