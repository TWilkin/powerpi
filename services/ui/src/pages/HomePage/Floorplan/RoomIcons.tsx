import { Room } from "@powerpi/common-api";
import { useCallback, useMemo } from "react";
import { useTranslation } from "react-i18next";
import { chain as _ } from "underscore";
import DeviceIcon from "../../../components/DeviceIcon";
import Icon from "../../../components/Icon";
import SensorIcon from "../../../components/SensorIcon";
import { isPolygonRoom } from "./ViewBox";
import useRoomDevices from "./useRoomDevices";

type RoomIconsProps = {
    room: Room;
};

/** Component to display device/sensor icons in a room on the floorplan. */
const RoomIcons = ({ room }: RoomIconsProps) => {
    const { t } = useTranslation();

    const devices = useRoomDevices(room.name);

    const { scale, iconPaddedSize, deviceCount, iconsWide, offsetX, offsetY } = useMemo(() => {
        const iconSize = 40;
        const iconPadding = 4;
        const iconPaddedSize = iconSize + 2 * iconPadding;

        const scale = 1 / (1024 / iconSize);

        // calculate the room size
        let x = room.x ?? 0;
        let y = room.y ?? 0;
        let width = 0;
        let height = 0;

        // this could put the icons outside the polygon so needs to be improved
        if (isPolygonRoom(room)) {
            x = Math.min(...room.points.map((point) => point.x));
            y = Math.min(...room.points.map((point) => point.y));

            width = Math.max(...room.points.map((point) => point.x)) - x;
            height = Math.max(...room.points.map((point) => point.y)) - y;
        } else {
            width = room.width ?? 0;
            height = room.height ?? 0;
        }

        // next we need to find the centre of the room
        const centreX = width / 2 + x;
        const centreY = height / 2 + y;

        // how many icons can we fit?
        let iconsWide = Math.floor(width / iconPaddedSize);
        let iconsTall = Math.floor(height / iconPaddedSize);

        // we want to leave some space either side
        iconsWide = Math.max(iconsWide - 2, 0);
        iconsTall = Math.max(iconsTall - 2, 0);

        // limit them to what is necessary to centre everything
        const best = bestValue(devices.length, iconsWide, iconsTall);
        if (best) {
            iconsWide = best[0];
            iconsTall = best[1];
        }

        // how many devices can we actually show
        let deviceCount = Math.min(devices.length, iconsWide * iconsTall);
        if (deviceCount < devices.length) {
            // if we have too many devices we'll show an ellipsis instead of the last one
            deviceCount--;
        }

        // now we can work out the offsets
        const offsetX = centreX - (iconsWide * iconPaddedSize) / 2;
        const offsetY = centreY - (iconsTall * iconPaddedSize) / 2;

        return {
            scale,
            iconPaddedSize,
            deviceCount,
            iconsWide,
            offsetX,
            offsetY,
        };
    }, [devices.length, room]);

    const getTransform = useCallback(
        (index: number) => {
            // find out which row and column it's in
            let row = 0;
            while (index >= (row + 1) * iconsWide) {
                row++;
            }
            const column = index - row * iconsWide;

            // now we can work out the position
            const x = offsetX + column * iconPaddedSize;
            const y = offsetY + row * iconPaddedSize;

            return `translate(${x} ${y}) scale(${scale})`;
        },
        [iconPaddedSize, iconsWide, offsetX, offsetY, scale],
    );

    return (
        <>
            {_(devices)
                .first(deviceCount)
                .map((device, i) => (
                    <g key={`${device.deviceType}_${device.type}`} transform={getTransform(i)}>
                        <title>
                            {device.type} ({device.count})
                        </title>

                        {device.deviceType === "device" && <DeviceIcon type={device.type} />}

                        {device.deviceType === "sensor" && (
                            <SensorIcon type={device.type} state={device.state} />
                        )}
                    </g>
                ))
                .value()}

            {deviceCount < devices.length && (
                <g transform={getTransform(deviceCount)}>
                    <title>
                        {t("pages.home.more devices", { count: devices.length - deviceCount })}
                    </title>

                    <Icon icon="more" />
                </g>
            )}
        </>
    );
};
export default RoomIcons;

/** Generate factors for value starting with the values closes to the square root. */
function* factors(value: number) {
    const range = Math.floor(Math.sqrt(value));

    for (let i = range; i >= 1; i--) {
        if (value % i === 0) {
            yield [i, value / i];
        }
    }
}

/** Generate factors between count and rows * columns, in case the factors of count won't fit. */
function* increasingFactors(count: number, rows: number, columns: number) {
    const maxCount = Math.min(count, rows * columns);

    for (let i = maxCount; i <= rows * columns; i++) {
        for (const f of factors(i)) {
            yield f;
        }
    }
}

/** Generate factors of count that will work for the rows and columns. */
function* feasibleFactors(count: number, rows: number, columns: number) {
    for (const [f1, f2] of increasingFactors(count, rows, columns)) {
        if (f1 <= columns && f2 <= rows) {
            yield [f2, f1];
        } else if (f2 <= columns && f1 <= rows) {
            yield [f1, f2];
        }
    }
}

/** Find the "best" factors of count in a grid of no more than rows x columns. */
function bestValue(count: number, rows: number, columns: number) {
    for (const factors of feasibleFactors(count, rows, columns)) {
        return factors;
    }

    return undefined;
}
