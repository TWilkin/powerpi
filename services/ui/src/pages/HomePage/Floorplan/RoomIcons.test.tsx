import { DeviceState } from "@powerpi/common-api";
import { render, screen } from "@testing-library/react";
import { vi } from "vitest";
import RoomIcons from "./RoomIcons";

const mocks = vi.hoisted(() => ({
    useQueryDevices: vi.fn(),
    userQuerySensors: vi.fn(),
}));

vi.mock("../../../queries/useQueryDevices", () => ({ default: mocks.useQueryDevices }));
vi.mock("../../../queries/useQuerySensors", () => ({ default: mocks.userQuerySensors }));

const devices = [
    createDevice("Socket1", "socket"),
    createDevice("Socket2", "socket"),
    createDevice("Light1", "light"),
    createDevice("Light2", "light"),
    createDevice("Light3", "light"),
];

const sensors = [
    createDevice("Temperature1", "temperature"),
    createDevice("Temperature2", "temperature"),
];

describe("RoomIcon", () => {
    beforeEach(() => {
        mocks.useQueryDevices.mockReturnValue({
            data: devices,
        });
        mocks.userQuerySensors.mockReturnValue({ data: sensors });
    });

    test("enough room", () => {
        const { container } = render(
            <svg>
                <RoomIcons
                    room={{
                        name: "Some Room",
                        width: 400,
                        height: 400,
                    }}
                />
            </svg>,
        );

        // there are 3 types
        expect(screen.getAllByRole("img", { hidden: true })).toHaveLength(3);

        expect(screen.getByText(/light \(3\)/)).toBeInTheDocument();
        expect(screen.getByText(/socket \(2\)/)).toBeInTheDocument();
        expect(screen.getByText(/temperature \(2\)/)).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test("too small", () => {
        const { container } = render(
            <svg>
                <RoomIcons
                    room={{
                        name: "Some Room",
                        width: 190,
                        height: 130,
                    }}
                />
            </svg>,
        );

        // we should see 1 icon and the ellipsis
        const icons = screen.getAllByRole("img", { hidden: true });
        expect(icons).toHaveLength(2);

        expect(icons[0]).toHaveAttribute("data-icon", "lightbulb");
        expect(icons[1]).toHaveAttribute("data-icon", "ellipsis");

        expect(screen.getByText(/There are 2 more hidden devices\/sensors./)).toBeInTheDocument();

        expect(screen.getByText(/light \(3\)/)).toBeInTheDocument();
        expect(screen.queryByText(/socket \(2\)/)).not.toBeInTheDocument();
        expect(screen.queryByText(/temperature \(2\)/)).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test("narrow", () => {
        const { container } = render(
            <svg>
                <RoomIcons
                    room={{
                        name: "Some Room",
                        x: 100,
                        y: 200,
                        width: 130,
                        height: 400,
                    }}
                />
            </svg>,
        );

        // there are 3 types
        expect(screen.getAllByRole("img", { hidden: true })).toHaveLength(3);

        expect(screen.getByText(/light \(3\)/)).toBeInTheDocument();
        expect(screen.getByText(/socket \(2\)/)).toBeInTheDocument();
        expect(screen.getByText(/temperature \(2\)/)).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test("polygon", () => {
        const { container } = render(
            <svg>
                <RoomIcons
                    room={{
                        name: "Some Room",
                        points: [
                            { x: 0, y: 0 },
                            { x: 500, y: 0 },
                            { x: 500, y: 600 },
                            { x: 0, y: 600 },
                        ],
                    }}
                />
            </svg>,
        );

        // there are 3 types
        expect(screen.getAllByRole("img", { hidden: true })).toHaveLength(3);

        expect(screen.getByText(/light \(3\)/)).toBeInTheDocument();
        expect(screen.getByText(/socket \(2\)/)).toBeInTheDocument();
        expect(screen.getByText(/temperature \(2\)/)).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
});

function createDevice(name: string, type: string, location: string = "Some Room", visible = true) {
    return {
        name,
        display_name: name,
        type,
        location,
        state: DeviceState.Unknown,
        since: 1,
        visible,
    };
}
