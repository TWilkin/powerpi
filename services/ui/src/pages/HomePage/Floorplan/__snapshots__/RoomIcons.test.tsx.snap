// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RoomIcon > enough room 1`] = `
<div>
  <svg>
    <g
      transform="translate(140 180) scale(0.03125)"
    >
      <title>
        light
         (
        3
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-lightbulb "
        data-icon="lightbulb"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2c0 0 0 0 0 0c5.2-7.1 10.4-14.2 15.4-21.4c19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3c5 7.2 10.2 14.3 15.4 21.4c0 0 0 0 0 0c19.8 27.1 39.7 54.4 49.2 86.2l160 0zM192 512c44.2 0 80-35.8 80-80l0-16-160 0 0 16c0 44.2 35.8 80 80 80zM112 176c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112c8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"
          fill="currentColor"
        />
      </svg>
    </g>
    <g
      transform="translate(180 180) scale(0.03125)"
    >
      <title>
        socket
         (
        2
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-plug "
        data-icon="plug"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 0C78.3 0 64 14.3 64 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM288 0c-17.7 0-32 14.3-32 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM32 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l0 32c0 77.4 55 142 128 156.8l0 67.2c0 17.7 14.3 32 32 32s32-14.3 32-32l0-67.2C297 398 352 333.4 352 256l0-32c17.7 0 32-14.3 32-32s-14.3-32-32-32L32 160z"
          fill="currentColor"
        />
      </svg>
    </g>
    <g
      transform="translate(220 180) scale(0.03125)"
    >
      <title>
        temperature
         (
        2
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-temperature-half "
        data-icon="temperature-half"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M160 64c-26.5 0-48 21.5-48 48l0 164.5c0 17.3-7.1 31.9-15.3 42.5C86.2 332.6 80 349.5 80 368c0 44.2 35.8 80 80 80s80-35.8 80-80c0-18.5-6.2-35.4-16.7-48.9c-8.2-10.6-15.3-25.2-15.3-42.5L208 112c0-26.5-21.5-48-48-48zM48 112C48 50.2 98.1 0 160 0s112 50.1 112 112l0 164.4c0 .1 .1 .3 .2 .6c.2 .6 .8 1.6 1.7 2.8c18.9 24.4 30.1 55 30.1 88.1c0 79.5-64.5 144-144 144S16 447.5 16 368c0-33.2 11.2-63.8 30.1-88.1c.9-1.2 1.5-2.2 1.7-2.8c.1-.3 .2-.5 .2-.6L48 112zM208 368c0 26.5-21.5 48-48 48s-48-21.5-48-48c0-20.9 13.4-38.7 32-45.3L144 208c0-8.8 7.2-16 16-16s16 7.2 16 16l0 114.7c18.6 6.6 32 24.4 32 45.3z"
          fill="currentColor"
        />
      </svg>
    </g>
  </svg>
</div>
`;

exports[`RoomIcon > narrow 1`] = `
<div>
  <svg>
    <g
      transform="translate(145 340) scale(0.03125)"
    >
      <title>
        light
         (
        3
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-lightbulb "
        data-icon="lightbulb"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2c0 0 0 0 0 0c5.2-7.1 10.4-14.2 15.4-21.4c19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3c5 7.2 10.2 14.3 15.4 21.4c0 0 0 0 0 0c19.8 27.1 39.7 54.4 49.2 86.2l160 0zM192 512c44.2 0 80-35.8 80-80l0-16-160 0 0 16c0 44.2 35.8 80 80 80zM112 176c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112c8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"
          fill="currentColor"
        />
      </svg>
    </g>
    <g
      transform="translate(145 380) scale(0.03125)"
    >
      <title>
        socket
         (
        2
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-plug "
        data-icon="plug"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 0C78.3 0 64 14.3 64 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM288 0c-17.7 0-32 14.3-32 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM32 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l0 32c0 77.4 55 142 128 156.8l0 67.2c0 17.7 14.3 32 32 32s32-14.3 32-32l0-67.2C297 398 352 333.4 352 256l0-32c17.7 0 32-14.3 32-32s-14.3-32-32-32L32 160z"
          fill="currentColor"
        />
      </svg>
    </g>
    <g
      transform="translate(145 420) scale(0.03125)"
    >
      <title>
        temperature
         (
        2
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-temperature-half "
        data-icon="temperature-half"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M160 64c-26.5 0-48 21.5-48 48l0 164.5c0 17.3-7.1 31.9-15.3 42.5C86.2 332.6 80 349.5 80 368c0 44.2 35.8 80 80 80s80-35.8 80-80c0-18.5-6.2-35.4-16.7-48.9c-8.2-10.6-15.3-25.2-15.3-42.5L208 112c0-26.5-21.5-48-48-48zM48 112C48 50.2 98.1 0 160 0s112 50.1 112 112l0 164.4c0 .1 .1 .3 .2 .6c.2 .6 .8 1.6 1.7 2.8c18.9 24.4 30.1 55 30.1 88.1c0 79.5-64.5 144-144 144S16 447.5 16 368c0-33.2 11.2-63.8 30.1-88.1c.9-1.2 1.5-2.2 1.7-2.8c.1-.3 .2-.5 .2-.6L48 112zM208 368c0 26.5-21.5 48-48 48s-48-21.5-48-48c0-20.9 13.4-38.7 32-45.3L144 208c0-8.8 7.2-16 16-16s16 7.2 16 16l0 114.7c18.6 6.6 32 24.4 32 45.3z"
          fill="currentColor"
        />
      </svg>
    </g>
  </svg>
</div>
`;

exports[`RoomIcon > polygon 1`] = `
<div>
  <svg>
    <g
      transform="translate(190 280) scale(0.03125)"
    >
      <title>
        light
         (
        3
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-lightbulb "
        data-icon="lightbulb"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2c0 0 0 0 0 0c5.2-7.1 10.4-14.2 15.4-21.4c19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3c5 7.2 10.2 14.3 15.4 21.4c0 0 0 0 0 0c19.8 27.1 39.7 54.4 49.2 86.2l160 0zM192 512c44.2 0 80-35.8 80-80l0-16-160 0 0 16c0 44.2 35.8 80 80 80zM112 176c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112c8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"
          fill="currentColor"
        />
      </svg>
    </g>
    <g
      transform="translate(230 280) scale(0.03125)"
    >
      <title>
        socket
         (
        2
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-plug "
        data-icon="plug"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 0C78.3 0 64 14.3 64 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM288 0c-17.7 0-32 14.3-32 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM32 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l0 32c0 77.4 55 142 128 156.8l0 67.2c0 17.7 14.3 32 32 32s32-14.3 32-32l0-67.2C297 398 352 333.4 352 256l0-32c17.7 0 32-14.3 32-32s-14.3-32-32-32L32 160z"
          fill="currentColor"
        />
      </svg>
    </g>
    <g
      transform="translate(270 280) scale(0.03125)"
    >
      <title>
        temperature
         (
        2
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-temperature-half "
        data-icon="temperature-half"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M160 64c-26.5 0-48 21.5-48 48l0 164.5c0 17.3-7.1 31.9-15.3 42.5C86.2 332.6 80 349.5 80 368c0 44.2 35.8 80 80 80s80-35.8 80-80c0-18.5-6.2-35.4-16.7-48.9c-8.2-10.6-15.3-25.2-15.3-42.5L208 112c0-26.5-21.5-48-48-48zM48 112C48 50.2 98.1 0 160 0s112 50.1 112 112l0 164.4c0 .1 .1 .3 .2 .6c.2 .6 .8 1.6 1.7 2.8c18.9 24.4 30.1 55 30.1 88.1c0 79.5-64.5 144-144 144S16 447.5 16 368c0-33.2 11.2-63.8 30.1-88.1c.9-1.2 1.5-2.2 1.7-2.8c.1-.3 .2-.5 .2-.6L48 112zM208 368c0 26.5-21.5 48-48 48s-48-21.5-48-48c0-20.9 13.4-38.7 32-45.3L144 208c0-8.8 7.2-16 16-16s16 7.2 16 16l0 114.7c18.6 6.6 32 24.4 32 45.3z"
          fill="currentColor"
        />
      </svg>
    </g>
  </svg>
</div>
`;

exports[`RoomIcon > too small 1`] = `
<div>
  <svg>
    <g
      transform="translate(55 45) scale(0.03125)"
    >
      <title>
        light
         (
        3
        )
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-lightbulb "
        data-icon="lightbulb"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2c0 0 0 0 0 0c5.2-7.1 10.4-14.2 15.4-21.4c19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3c5 7.2 10.2 14.3 15.4 21.4c0 0 0 0 0 0c19.8 27.1 39.7 54.4 49.2 86.2l160 0zM192 512c44.2 0 80-35.8 80-80l0-16-160 0 0 16c0 44.2 35.8 80 80 80zM112 176c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112c8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"
          fill="currentColor"
        />
      </svg>
    </g>
    <g
      transform="translate(95 45) scale(0.03125)"
    >
      <title>
        There are 2 more hidden devices/sensors.
      </title>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-ellipsis "
        data-icon="ellipsis"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
          fill="currentColor"
        />
      </svg>
    </g>
  </svg>
</div>
`;
